File: mobile/lib/domain/ai/coach_inbox.dart
// In-App-Postfach — Spiegel src/lib/ai/coachInbox.ts.

import 'coach_watch.dart';

const coachSnoozeDays = 7;

class CoachMeta {
  const CoachMeta({this.snoozedUntil, this.readFingerprint});

  final DateTime? snoozedUntil;
  final String? readFingerprint;

  Map<String, dynamic> toJson() => {
        if (snoozedUntil != null)
          'snoozedUntil': snoozedUntil!.toUtc().toIso8601String(),
        if (readFingerprint != null) 'readFingerprint': readFingerprint,
      };

  factory CoachMeta.fromJson(Map<String, dynamic> json) {
    DateTime? until;
    final raw = json['snoozedUntil'] as String?;
    if (raw != null) until = DateTime.tryParse(raw)?.toUtc();
    return CoachMeta(
      snoozedUntil: until,
      readFingerprint: json['readFingerprint'] as String?,
    );
  }

  CoachMeta copyWith({
    DateTime? snoozedUntil,
    String? readFingerprint,
  }) {
    return CoachMeta(
      snoozedUntil: snoozedUntil ?? this.snoozedUntil,
      readFingerprint: readFingerprint ?? this.readFingerprint,
    );
  }
}

class CoachInboxItem {
  const CoachInboxItem({required this.notice, required this.unread});
  final CoachNotice notice;
  final bool unread;
}

bool isSnoozed(CoachMeta? meta, DateTime now) {
  final until = meta?.snoozedUntil;
  if (until == null) return false;
  return until.isAfter(now);
}

List<CoachInboxItem> mergeCoachInbox(
  List<CoachNotice> notices,
  Map<String, CoachMeta> meta, {
  DateTime? now,
}) {
  final clock = now ?? DateTime.now().toUtc();
  final items = <CoachInboxItem>[];
  for (final n in notices) {
    final prev = meta[n.id];
    if (isSnoozed(prev, clock)) continue;
    final unread =
        prev?.readFingerprint == null || prev!.readFingerprint != n.fingerprint;
    items.add(CoachInboxItem(notice: n, unread: unread));
  }
  return items;
}

Map<String, CoachMeta> snoozeMeta(
  Map<String, CoachMeta> meta,
  CoachNotice notice, {
  int days = coachSnoozeDays,
  DateTime? now,
}) {
  final clock = now ?? DateTime.now().toUtc();
  final next = Map<String, CoachMeta>.from(meta);
  next[notice.id] = CoachMeta(
    snoozedUntil: clock.add(Duration(days: days)),
    readFingerprint: notice.fingerprint,
  );
  return next;
}

Map<String, CoachMeta> markReadMeta(
  Map<String, CoachMeta> meta,
  Iterable<CoachNotice> notices,
) {
  final next = Map<String, CoachMeta>.from(meta);
  for (final n in notices) {
    final prev = next[n.id];
    next[n.id] = CoachMeta(
      snoozedUntil: prev?.snoozedUntil,
      readFingerprint: n.fingerprint,
    );
  }
  return next;
}

int unreadCoachCount(List<CoachInboxItem> items) =>
    items.where((i) => i.unread).length;

File: src/lib/ai/coachInbox.ts
/**
 * In-App-Postfach für Coach-Hinweise.
 * Snooze 7 Tage; bei geändertem Fingerprint (z. B. bald → überfällig) wieder ungelesen.
 */

import type { CoachNotice } from "@/lib/ai/coachWatch";

export const COACH_SNOOZE_DAYS = 7;

export interface CoachMeta {
  snoozedUntil?: string;
  readFingerprint?: string;
}

export interface CoachInboxItem extends CoachNotice {
  unread: boolean;
}

export function isSnoozed(meta: CoachMeta | undefined, now: Date): boolean {
  if (!meta?.snoozedUntil) return false;
  const t = Date.parse(meta.snoozedUntil);
  return Number.isFinite(t) && t > now.getTime();
}

export function mergeCoachInbox(
  notices: CoachNotice[],
  meta: Record<string, CoachMeta>,
  now = new Date()
): CoachInboxItem[] {
  const items: CoachInboxItem[] = [];
  for (const n of notices) {
    const prev = meta[n.id];
    if (isSnoozed(prev, now)) continue;
    const unread = !prev?.readFingerprint || prev.readFingerprint !== n.fingerprint;
    items.push({ ...n, unread });
  }
  return items;
}

export function snoozeMeta(
  meta: Record<string, CoachMeta>,
  notice: Pick<CoachNotice, "id" | "fingerprint">,
  days = COACH_SNOOZE_DAYS,
  now = new Date()
): Record<string, CoachMeta> {
  const until = new Date(now.getTime() + days * 24 * 60 * 60 * 1000);
  return {
    ...meta,
    [notice.id]: {
      ...meta[notice.id],
      snoozedUntil: until.toISOString(),
      readFingerprint: notice.fingerprint,
    },
  };
}

export function markReadMeta(
  meta: Record<string, CoachMeta>,
  notices: Pick<CoachNotice, "id" | "fingerprint">[]
): Record<string, CoachMeta> {
  const next = { ...meta };
  for (const n of notices) {
    next[n.id] = {
      ...next[n.id],
      readFingerprint: n.fingerprint,
    };
  }
  return next;
}

export function unreadCoachCount(items: CoachInboxItem[]): number {
  return items.filter((i) => i.unread).length;
}

export function pruneCoachMeta(
  meta: Record<string, CoachMeta>,
  notices: CoachNotice[]
): Record<string, CoachMeta> {
  const keep = new Set(notices.map((n) => n.id));
  const next: Record<string, CoachMeta> = {};
  for (const [id, row] of Object.entries(meta)) {
    if (keep.has(id) || row.snoozedUntil) next[id] = row;
  }
  return next;
}
